security.md

```diff
@@ -0,0 +1,94 @@
+even successful attacks can be detected.
+
+In general, this is the same threat model as with ordinary key exchanges, except
+that attackers additionally have to guess the code correctly. (Hence the name,
+"*password authenticated* key exchange", PAKE.)
+
+### Password entropy
+
+Compared to password strengths we are used to nowadays, 16 bits may seem scarily
+small, even knowing that an attacker only has one try to guess the password. But
+to put it into perspective: As a user, you'd have to experience 32000 failed
+transfers because of an attack in order to having a 50% chance of a file being
+intercepted. From an attacker's perspective, guessing thousands of codes is not
+difficult, but there is no possibility of targetting any individual connection,
+making any gained data fairly useless. From a service perspective, any successful
+attack implies 65536 failed attempts on average, which will quickly be noticed
+due to their disrupting nature.
+
+In short, guessing codes is unlikely to be an attractive target for any attacker.
+Of course, there is always the possibility to configure a longer code in the
+clients if desired.
+
+## Denial of Service (DoS) attacks on the Rendezvous server
+
+Wormhole codes can be so short because they implicitly contain a common
+rendezvous server URL (any two applications that use Magic Wormhole
+should be configured to use the same server). As a result, successful
+operation depends upon both clients being able to contact that server,
+making it a single point of failure (SPOF).
+
+While it is of course possible to incapacitate the rendezvous server using
+traditional means of DoS, so much work is not necessary: An attacker may
+simply connect to every nameplate with a random code, causing the key exchange
+to fail. There is a "list" command in the protocol which makes it easy to
+enumerate all nameplates to disrupt them, but even without it the name space is
+sufficiently small (by design, as we want short codes) to brute force.
+
+There are several ideas to defend against this, however the only one already
+specified in the protocol is the "permission" feature: The rendezvous server
+may use it to restrict new connections with some kind of challenge (or for
+private servers, authentication). The motivation behind this is to enable
+a proof-of-work challenge (using HashCash) when under attack to slow it down.
+See the server protocol specification for details.
+
+Currently the best defense against this is to use high-entropy nameplates which
+cannot easily be enumerated at the cost of longer codes. This has not be
+implemented yet though, and would still leave the attack vector of "ordinary"
+DoS.
+
+## Metadata leaks
+
+As usual in cryptography, communication metadata cannot easily be protected. Any
+involved server will know the timestamp and size of each message it relays. This
+usually is not much of an issue, except for file transfer where the file size may
```

Comment:
Well, _really_ depends on perspective? (What's a case where it isn't an issue?)

Reply:
Most of the time actually, the file size does not contain that much information: an attacker gains nothing of value from the fact that my configuration file has 45219 bytes (ignoring for now that the protocol may round numbers up to some block size for encryption).

Reply:
File-sizes are actually extremely unique (unfortunately). This e.g. forms part of the basis of "fingerprinting" Web pages, which is extremely successful (e.g. allowing one to answer the question "what website is this TLS connection visiting" even on networks like Tor).

So, I would assume this can be extended to pattern-match on message flows in wormhole as well (allowing fingerprinting of any protocol). I guess it's the "not much of an issue" comment that I'm commenting on here...

Reply:
Trying to be more-clear: leaking metadata is bad.

So, it's still bad for almost any "application protocol" I can think of, and it should remain clear that Wormhole does NOT protect against that -- allowing passive observers (including the mailbox server) to learn many things about your protocol even though the *contents* remain encrypted an unview-able by the server.